Firestore/core/src/firebase/firestore/core/firestore_client.mm

```diff
@@ -0,0 +1,485 @@
+  // before that subsequent work completes.
+  if (settings.persistence_enabled()) {
+    Path dir = [FSTLevelDB
+        storageDirectoryForDatabaseInfo:database_info_
+                     documentsDirectory:[FSTLevelDB documentsDirectory]];
+
+    FSTSerializerBeta* remote_serializer = [[FSTSerializerBeta alloc]
+        initWithDatabaseID:database_info_.database_id()];
+    FSTLocalSerializer* serializer =
+        [[FSTLocalSerializer alloc] initWithRemoteSerializer:remote_serializer];
+    FSTLevelDB* ldb;
+    Status level_db_status = [FSTLevelDB
+        dbWithDirectory:std::move(dir)
+             serializer:serializer
+              lruParams:LruParams::WithCacheSize(settings.cache_size_bytes())
+                    ptr:&ldb];
+    if (!level_db_status.ok()) {
+      // If leveldb fails to start then just throw up our hands: the error is
+      // unrecoverable. There's nothing an end-user can do and nearly all
+      // failures indicate the developer is doing something grossly wrong so we
+      // should stop them cold in their tracks with a failure they can't ignore.
+      [NSException
+           raise:NSInternalInconsistencyException
+          format:@"Failed to open DB: %s", level_db_status.ToString().c_str()];
+    }
+    lru_delegate_ = ldb.referenceDelegate;
+    persistence_ = ldb;
+    if (settings.gc_enabled()) {
+      ScheduleLruGarbageCollection();
+    }
+  } else {
+    persistence_ = [FSTMemoryPersistence persistenceWithEagerGC];
+  }
+
+  local_store_ = [[FSTLocalStore alloc] initWithPersistence:persistence_
+                                                initialUser:user];
+
+  auto datastore = std::make_shared<Datastore>(database_info_, worker_queue(),
+                                               credentials_provider_);
+
+  auto shared_self = shared_from_this();
+  remote_store_ = absl::make_unique<RemoteStore>(
+      local_store_, std::move(datastore), worker_queue(),
+      [shared_self](OnlineState online_state) {
+        [shared_self->sync_engine_ applyChangedOnlineState:online_state];
+      });
+
+  sync_engine_ = [[FSTSyncEngine alloc] initWithLocalStore:local_store_
+                                               remoteStore:remote_store_.get()
+                                               initialUser:user];
+
+  event_manager_.Init(sync_engine_);
+
+  // Setup wiring for remote store.
+  remote_store_->set_sync_engine(sync_engine_);
+
+  // NOTE: RemoteStore depends on LocalStore (for persisting stream tokens,
+  // refilling mutation queue, etc.) so must be started after LocalStore.
+  [local_store_ start];
+  remote_store_->Start();
+}
+
+/**
+ * Schedules a callback to try running LRU garbage collection. Reschedules
+ * itself after the GC has run.
+ */
+void FirestoreClient::ScheduleLruGarbageCollection() {
+  std::chrono::milliseconds delay =
+      gc_has_run_ ? regular_gc_delay_ : initial_gc_delay_;
+  auto shared_self = shared_from_this();
```

Comment:
@wilhuff What's your take on using a shared versus raw pointer when putting something on the worker queue from this class?

Previously, when `FirestoreClient` owned the queue, it was clear that raw pointer was sufficient. The fact that queue ownership is shared seems to mean the exact opposite, that the queue might outlive the client (otherwise, there would be no need for shared ownership). My advice was to use shared pointer for two reasons:
* I'm not clear on whether the ownership problem may happen or not, and it seems like something infeasible to prove one way or another and also not enforced (so even if it is not a problem now, nothing prevents it from becoming a problem in the future);
* if the problem actually exists, it would probably appear very rarely and be pretty hard to debug.

Having said that, I don't have the full context on how the worker queue is used, so I'd like to hear your take on this.

Reply:
This seems pretty harmless and is markedly safer than trying to reason about lifetimes, which we've gotten wrong publicly several times now.

I'm in favor.